File: kernel-sidecar/src/rayito_kernel_sidecar/logging.py
"""JSON lines on stderr, one object per line: ``{"level", "msg", <fields>}``.

``rayd`` parses these lines and re-emits only the allowlisted fields through
its own logger, so a field outside ``ALLOWED_FIELDS`` is dropped here first.
``msg`` values are fixed literals; nothing that comes from a cell (code,
output, mime payloads, envs, cwd, tracebacks) ever becomes a field value:
lengths and counts travel instead (``code_len``, ``text_len``, ``chunks``).
"""

from __future__ import annotations

import json
import sys
from typing import Any, Final, TextIO

ALLOWED_FIELDS: Final = frozenset(
    {
        "op",
        "id",
        "context_id",
        "execution_id",
        "execution_count",
        "events",
        "results",
        "mime_types",
        "timeout_ms",
        "outcome",
        "kernel_pid",
        "attempt",
        "backoff_ms",
        "exit_code",
        "warmup_ms",
        "restart_ms",
        "duration_ms",
        "bytes",
        "chunks",
        "contexts",
        "reseeded",
        "deferred",
        "failed",
        "skipped",
        "language",
        "languages",
        "code_len",
        "text_len",
        "reason",
        "state",
        "queued",
    }
)
LEVELS: Final = ("debug", "info", "warn", "error")


class SidecarLogger:
    def __init__(self, stream: TextIO | None = None) -> None:
        self._stream = stream if stream is not None else sys.stderr

    def log(self, level: str, msg: str, **fields: Any) -> None:
        if level not in LEVELS:
            level = "info"
        record: dict[str, Any] = {"level": level, "msg": msg}
        for key, value in fields.items():
            if key in ALLOWED_FIELDS and value is not None:
                record[key] = _plain(value)
        self._stream.write(json.dumps(record, separators=(",", ":"), ensure_ascii=False) + "\n")
        self._stream.flush()

    def debug(self, msg: str, **fields: Any) -> None:
        self.log("debug", msg, **fields)

    def info(self, msg: str, **fields: Any) -> None:
        self.log("info", msg, **fields)

    def warn(self, msg: str, **fields: Any) -> None:
        self.log("warn", msg, **fields)

    def error(self, msg: str, **fields: Any) -> None:
        self.log("error", msg, **fields)


def _plain(value: Any) -> Any:
    if isinstance(value, bool | int | float | str):
        return value
    if isinstance(value, list | tuple):
        return [_plain(item) for item in value]
    return str(value)

File: crates/rayd/src/logging.rs
//! JSON lines on stderr, filtered by `RAYD_LOG` (default `info`).
//!
//! Field allowlist for every event in this crate: `hook`, `rpc`, `phase`,
//! `from`, `to`, `outcome`, `reason`, `sandbox_id`, `suspend_generation`,
//! `resume_generation`, `clock_offset_ms`, `payload_chars`, `budget_ms`,
//! `grpc_port`, `hooks_port`, `version`, and for processes `pid`, `status`,
//! `exit_code`, `signal`, `seq`, `step`, `subscribers`, `live_processes`,
//! `reaped`, `duration_ms`, `errno`, `identity_switch`, `nofile_soft`,
//! `nofile_hard`, and for files `files`, `bytes`, `chunks`, `entries`,
//! `depth`, `recursive`, `events`, `live_watches`, `timeout_ms`,
//! `denied_prefixes`, and for code execution `context_id`, `execution_id`,
//! `execution_count`, `results`, `mime_types` (names only), `outcome`,
//! `kernel_pid`, `attempt`, `backoff_ms`, `warmup_ms`, `restart_ms`,
//! `restart_s`, `sidecar_restarts`, `sidecar_stderr_lines`, `contexts`,
//! `reseeded`, `failed`, `pending`, `kernels_killed`, `dropped_events`,
//! `consecutive_timeouts`, `state`, `source`, `msg` and `fields` (the
//! sidecar's own allowlisted fields as compact JSON), and for terminals
//! and suspend/resume `cols`, `rows`, `pty_devices`, `suspended_ms`,
//! `streams_closed`, `streams_pending`, `suspend_ms`, `probe_ms`,
//! `kernels_alive`, `kernels_lost`, `kernel_state_lost`, `reattach`,
//! `replayed`, `from_seq`, `op_timeout_across_resume`, and for the M6
//! hardening `hook_audit` (a message), `calls_since_run`, `anomaly`,
//! `hook_anomalies`, `stale_suspend_recovered`,
//! `resume_after_stale_recovery`, `capabilities`, `net_admin`,
//! `sys_admin`, `sys_resource`, `sys_ptrace`, `cgroup2_root`, `imds_probe`,
//! `imds_blocked`, `rule_present`, `root_reachable`, `user_probe_ran`,
//! `user_reachable`, `imds_block_unavailable`, `imds_rule_missing`, `cpu_seconds`,
//! `metadata_keys` (a count), `output_budget_bytes`, `entries_dropped`,
//! `disk_reserve`, `disk_full`, `free_bytes`, `deferred`,
//! `advisory_op_timeout`, and for the logical deadline (ADR-011)
//! `sandbox_timeout` and `timeout_forced_exit` (messages), `on_timeout`,
//! `extensions`, `overrun_ms`, `action`, `mode` and `lifecycle_phase`. Request bodies,
//! payloads, tokens, digests, commands, the shell's arguments,
//! environments, working directories, tags, stdin bytes, output bytes,
//! terminal bytes (input or output), paths, entry names, symlink targets,
//! file contents, chunks, code, cell output, mime payloads, error values,
//! traceback lines, connection files, the sidecar's raw stderr, metadata
//! keys or values, `ip` output and the IMDS probes' output are
//! never logged; error `Display` strings used as `reason` are written so
//! they never quote input.

use tracing_subscriber::EnvFilter;

pub const LOG_FILTER_ENV: &str = "RAYD_LOG";

pub fn init() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let filter = EnvFilter::try_from_env(LOG_FILTER_ENV).unwrap_or_else(|_| EnvFilter::new("info"));
    tracing_subscriber::fmt()
        .json()
        .with_env_filter(filter)
        .with_writer(std::io::stderr)
        .with_target(false)
        .try_init()
}
